File: Estructurales/facade.py

class VerificadorTarjeta:
    def verificar(self, tarjeta, cvv):
        print(f"Verificando tarjeta {tarjeta} con CVV {cvv}...")
        return True
    
class ProcesadorPago:
    def procesar(self, tarjeta, monto):
        print(f"Procesando pago de ${monto} con tarjeta {tarjeta}...")
        return True
    
class Notificador:
    def enviar_confirmacion(self, usuario):
        print(f"Enviando confirmación de pago a {usuario}...")
        
class PagoFacade:
    def __init__(self):
        self.verificador = VerificadorTarjeta()
        self.procesador = ProcesadorPago()
        self.notificador = Notificador()
        
    def realizar_pago(self, usuario, tarjeta, cvv, monto):
        print("Iniciando proceso de pago...")
        if not self.verificador.verificar(tarjeta, cvv):
            print("Pago Fallido: tarjeta no valida")
            return False
        if not self.procesador.procesar(tarjeta, monto):
            print("Pago fallido: error en el procesamiento")
            return False
        self.notificador.enviar_confirmacion(usuario)
        print("Pago realizado con exito")
        return True
    
pago= PagoFacade()
pago.realizar_pago("Gabriel Proaño", "1569-6718-2366-5432", "123", 100)

File: Ejercicios/Facade.java
package Ejercicios;

class VerificadorTarjeta {
    public boolean verificar(String tarjeta, String cvv) {
        System.out.println("Verificando tarjeta " + tarjeta + " con CVV " + cvv);
        return true; 
    }
}

class ProcesadorPago {
    public boolean procesar(String tarjeta, double monto) {
        System.out.println("Procesando pago de $" + monto + " con tarjeta " + tarjeta);
        return true; 
    }
}

class Notificador {
    public void enviarConfirmacion(String usuario) {
        System.out.println("Enviando confirmación de pago a " + usuario);
    }
}


class PagoFacade {
    private VerificadorTarjeta verificador;
    private ProcesadorPago procesador;
    private Notificador notificador;

    public PagoFacade() {
        verificador = new VerificadorTarjeta();
        procesador = new ProcesadorPago();
        notificador = new Notificador();
    }

    public boolean realizarPago(String usuario, String tarjeta, String cvv, double monto) {
        System.out.println("Iniciando proceso de pago...");
        if (!verificador.verificar(tarjeta, cvv)) {
            System.out.println("Pago fallido: tarjeta no válida");
            return false;
        }
        if (!procesador.procesar(tarjeta, monto)) {
            System.out.println("Pago fallido: error en procesamiento");
            return false;
        }
        notificador.enviarConfirmacion(usuario);
        System.out.println("Pago realizado exitosamente");
        return true;
    }
}


public class Facade {
    public static void main(String[] args) {
        PagoFacade pago = new PagoFacade();
        pago.realizarPago("Gabriel Proaño", "1569-6718-2366-5432", "123", 100);
    }
}